File: test_VH/sample1_worker2.f90
program sample_worker
  use mpi
  use ctca
  implicit none

  integer(kind=4) :: dataintnum, myrank, nprocs, ierr, areaid, progid, fromrank
  integer(kind=4) :: world_myrank
  integer(kind=4),dimension(2) :: dataint
  real(kind=8),dimension(6,400) :: datareal8
  character(len=256) :: fname
  integer(kind=4) :: i, j, c
  integer(kind=8) :: s
  s = 6*400
  
  call CTCAW_init(2, 2)
  print *, "worker2 init done"

  call MPI_Comm_size(CTCA_subcomm, nprocs, ierr)
  call MPI_Comm_rank(CTCA_subcomm, myrank, ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD, world_myrank, ierr)

  do while (.true.)
     call CTCAW_pollreq_withreal8(fromrank, dataint, 2, datareal8, s)
     c = dataint(2)

     if (CTCAW_isfin()) then
        exit
     end if

     ! set up data
     call MPI_Bcast(datareal8, 6*400, MPI_DOUBLE_PRECISION, 0, CTCA_subcomm, ierr)

     ! do work
     write(fname, '("out2.", i3.3,"-", i3.3, ".txt")') c, myrank
     open (10, file=fname, status="replace")
     do i = 1, 400
        write(10, *) datareal8(1:6, i)
     end do
     close(10)

     ! end this work
     call CTCAW_complete()
     print *, world_myrank, "worker2 complete done"

  end do

  call CTCAW_finalize()

end program sample_worker

File: test_VH/sample1_worker2.c
#include <stdio.h>
#include <mpi.h>
#include "ctca.h"

//#define N (1024UL*1024*1024)
#define N 1024

int main()
{
    int myrank, nprocs, progid, fromrank;
    int intparams[2];
    double data[N];
    int i, j, k, c;
    int prognum = 2;

    printf("worker2 init\n");
    CTCAW_init(1, 2);
    printf("worker2 init done\n");

    MPI_Comm_size(CTCA_subcomm, &nprocs);
    MPI_Comm_rank(CTCA_subcomm, &myrank);

    CTCAW_prof_start();
    CTCAW_prof_start_calc();

    while(1) {
        CTCAW_pollreq_withreal8(&fromrank, intparams, 2, data, N);
        c = intparams[1];

        if (CTCAW_isfin()) 
            break;

        printf("worker2 poll req done\n");

//        MPI_Bcast(data, 6*400, MPI_DOUBLE, 0, CTCA_subcomm);

//        printf("worker2 bcast done\n");

        printf("worker2 rank %d c %d data %f %f %f ... %f %f\n", myrank, c, data[0], data[1], data[2], data[N-2], data[N-1]);
        
        CTCAW_complete();
        printf("worker2 complete done\n");
    }

    CTCAW_prof_stop_calc();
    CTCAW_prof_stop();

    fprintf(stderr, "%d: worker2 finalize\n", myrank);
    CTCAW_finalize();
}
